ssg/rules.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import absolute_import
+from __future__ import print_function
+
+import os
+
+
+from .build_remediations import REMEDIATION_MAP
+
+
+def get_rule_dir_yaml(dir_path):
+    """
+    Returns the path to the yaml metadata for a rule directory,
+    regardless of if it exists.
+    """
+    return os.path.join(dir_path, "rule.yml")
+
+
+def get_rule_dir_id(path):
+    """
+    Returns the ID of a rule directory; correctly handles being passed
+    either the directory path or the yaml metadata path.
+    """
+    dir_path = path
+
+    if path.endswith("rule.yml"):
+        dir_path = os.path.dirname(path)
+
+    return os.path.basename(dir_path)
+
+
+def is_rule_dir(dir_path):
+    """
+    Returns True iff dir_path is a valid rule directory which exists
+
+    To be valid, dir_path must exist and be a directory and the file
+    returned by get_rule_dir_yaml(dir_path) must exist.
+    """
+    rule_yaml = get_rule_dir_yaml(dir_path)
+
+    is_dir = os.path.isdir(dir_path)
+    has_rule_yaml = os.path.exists(rule_yaml)
+
+    return is_dir and has_rule_yaml
+
+
+def get_rule_dir_ovals(dir_path, product=None):
```

Comment:
This function does two quite different things depending on whether product is None. What about having a private function with common part and one function without the product argument and another one that accepts it? Do we have a use case for a call with a product that is None?

Reply:
@matejak Yes, the use case is here:

https://github.com/cipherboy/scap-security-guide/blob/rule-dirs-tools/utils/rule_dir_json.py#L94

I would argue that it isn't that it does two _different_ things based on the value of `product`, but more that `product` is a filter which limits the output of `get_rule_dir_ovals` to only things which _could_ apply to that value of `product`, where `None` represents `all`.

The code could be improved by forward porting the [`is_applicable` function](https://github.com/cipherboy/scap-security-guide/blob/rule-dirs-tools/ssg/rules.py#L10-L20) (+ dependencies to parse the `<platform>` and configuration in remediations) introduced in that same `rule-dirs-tools` branch (for #3193) and having `is_applicable` return True `if product == None`. While that wouldn't _technically_ affect the build system (as the pruning of OVALs / remediations happen at a later time), it'd view it as a semantic change and in the case of `remediations`, requires the introduction of parsing `jinja`, which is best done with the entire build config (so we can use the jinja bytecode cache).

However, that would break the later use case in `rule-dirs-tools` and complicate the function signature. At that point, you might as well return the parsed OVAL/fix and that'd also be discarded for the tools. 

Thoughts?

Reply:
Yep, that's my current implementation. I'll push changes soon. :)